### PR TITLE
The AddOn

### PR DESCRIPTION
**Pull request by formerandroider**

### CODE REVIEW

**Comment by Xon** [comment]
@liamwli from what I can see, the usergroup isn't included when mentioned by a group. This has ended up being highly desirable functionality.

**Comment by formerandroider** [comment]
@Xon you mean show the user group title in the alert when mentioned with a user group? I'll have a look. I went this way because it significantly reduced the amount of class extensions.

**Comment by Xon** [comment]
Yes. The current alert text adds "You were tagged because you are a member of the {group} usergroup" to the standard XF mention text.

**Comment by formerandroider** [comment]
Ok. I'll add them back in.

File: Repository/UserMentions.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	 */
+	public function getMembersOfUserGroup(UserGroup $userGroup)
+	{
+		$memberUserIds = \XF::db()->fetchAllColumn('
+			SELECT DISTINCT user.user_id
+			FROM xf_user AS user
+			LEFT JOIN xf_user_group_relation AS relation ON relation.user_id = user.user_id
+            WHERE relation.user_group_id = ?', $userGroup->user_group_id);
+
+		$users = $this->finder('XF:User')->whereIds($memberUserIds)->fetch();
```
Comment (Xon) [inline Repository/UserMentions.php]: Can this be rewritten as a single query by extending UserGroup to have a relationship with User via xf_user_group_relation?

In the future I would like this pagable which is much easier if this is a standard XF finder relationship.

File: XF/Pub/View/Member/Find.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+class Find extends XFCP_Find
+{
+	public function renderJson()
+	{
+		$response = parent::renderJson();
+
+		foreach ($this->params['usergroups'] AS $usergroup)
+		{
+			$response['results'][] = [
```
Comment (Xon) [inline XF/Pub/View/Member/Find.php]: This is adding usergroups to the end of the list, They should be added to the start of the list to preserve ordering of preferring usergroup mention's 1st.

File: XF/Service/StructuredText/Preparer.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		', 'user_id');
+
+		if (!$additionalUsers)
+		{
+			return $users;
+		}
+
+		foreach ($additionalUsers AS $userId => $additionalUser)
+		{
+			$additionalUsers[$userId]['lower'] = strtolower($additionalUser['username']);
```
Comment (Xon) [inline XF/Service/StructuredText/Preparer.php]: Any reason to not just set $user rather than $additionalUsers and save an array concatenate?

Comment (formerandroider) [reply XF/Service/StructuredText/Preparer.php]: Nope, that'd work perfectly.

File: Repository/UserMentions.php (hunk 1 above)
Comment (formerandroider) [reply Repository/UserMentions.php]: I think I've managed this in 9c1fad5... I had to create our own UserGroupRelation entity though.

File: XF/Service/StructuredText/Preparer.php (hunk 3 above)
Comment (formerandroider) [reply XF/Service/StructuredText/Preparer.php]: Did this in de00501.

File: XF/Pub/View/Member/Find.php (hunk 2 above)
Comment (formerandroider) [reply XF/Pub/View/Member/Find.php]: Done in 082952d.

**Comment by formerandroider** [comment]
@Xon cracked adding the user group to the alert :)

File: XF/Repository/UserAlert.php [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+class UserAlert extends XFCP_UserAlert
+{
+	public function alert(
+		\XF\Entity\User $receiver, $senderId, $senderName,
+		$contentType, $contentId, $action, array $extra = []
+	)
+	{
+		if ($this->app()
+				->offsetExists('sv_userGroupMentionedIds') && ($contentType == 'post' || $contentType == 'profile_post') && $action == 'mention' && array_key_exists($receiver->user_id, $this->app()->sv_userGroupMentionedIds)
```
Comment (Xon) [inline XF/Repository/UserAlert.php]: I don't think it make sense for the explicit content type check, as without it; it allows easily extending to other content types (ie reports) without needing to edit to extend this code.

Comment (formerandroider) [reply XF/Repository/UserAlert.php]: Yeah, that makes sense.

**Comment by Xon** [comment]
I just realized that XF2 supports user mentions with report comments by default, can you extend those to support user groups?

Ideally this should be as low touch as possible to support other content types

**Comment by formerandroider** [comment]
@Xon the reports system uses the message preparer, which is the same as threads and posts, so it was already working - I just had to template mod the alert to add the extra text :)

I also moved some repeated code into the Repository class.

**Pull request merged**